firebase-dataconnect/firebase-dataconnect.gradle.kts

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.jetbrains.dokka.gradle.DokkaTask
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import io.javalin.Javalin
+import io.javalin.http.staticfiles.Location
+
+plugins {
+  id("firebase-library")
+  id("kotlin-android")
+  id("com.google.protobuf")
+  id("copy-google-services")
+  alias(libs.plugins.dokka)
+  alias(libs.plugins.kotlinx.serialization)
+}
+
+firebaseLibrary {
+  libraryGroup("dataconnect")
+  testLab.enabled = false
+  publishSources = true
+  publishJavadoc = false
+  previewMode = "alpha"
+}
+
+android {
+  val compileSdkVersion : Int by rootProject
+  val targetSdkVersion : Int by rootProject
+  val minSdkVersion : Int by rootProject
+
+  namespace = "com.google.firebase.dataconnect"
+  compileSdk = compileSdkVersion
+  defaultConfig {
+    minSdk = minSdkVersion
+    targetSdk = targetSdkVersion
+    multiDexEnabled = true
+    testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+  }
+  compileOptions {
+    sourceCompatibility = JavaVersion.VERSION_1_8
+    targetCompatibility = JavaVersion.VERSION_1_8
+  }
+  kotlinOptions { jvmTarget = "1.8" }
+
+  @Suppress("UnstableApiUsage")
+  testOptions {
+    unitTests {
+      isIncludeAndroidResources = true
+      isReturnDefaultValues = true
+    }
+  }
+
+  packaging {
+    resources {
+      excludes.add("META-INF/LICENSE.md")
+      excludes.add("META-INF/LICENSE-notice.md")
+    }
+  }
+}
+
+protobuf {
+  protoc {
+    artifact = "${libs.protoc.get()}"
+  }
+  plugins {
+    create("java") {
+      artifact = "${libs.grpc.protoc.gen.java.get()}"
+    }
+    create("grpc") {
+      artifact = "${libs.grpc.protoc.gen.java.get()}"
+    }
+    create("grpckt") {
+      artifact = "${libs.grpc.protoc.gen.kotlin.get()}:jdk8@jar"
+    }
+  }
+  generateProtoTasks {
+    all().forEach { task ->
+      task.builtins {
+        create("kotlin") {
+          option("lite")
+        }
+      }
+      task.plugins {
+        create("java") {
+          option("lite")
+        }
+        create("grpc") {
+          option("lite")
+        }
+        create("grpckt") {
+          option("lite")
+        }
+      }
+    }
+  }
+}
+
+dependencies {
+  api("com.google.firebase:firebase-common:21.0.0")
+
+  api(libs.kotlinx.coroutines.core)
+  api(libs.kotlinx.serialization.core)
```

Comment:
I feel like this discussion has been had before, but I couldn't find a record of it.

Technically, they don't _need_ to be `api`. For example, we don't expose coroutines on the consumers classpath for `firebase-database`; even though we have methods that return/accept `Flow`. This works, because consumers are expected to have coroutines on their classpath anyways- as it's a requirement on the Kotlin side. This also avoids the issue of adding additional bloat to the SDK size for consumers that _don't_ use those APIs, since coroutines is a "bulky" library.

BUT, we _should_ do this. The guidance is- if it's public api, then `api` it.

Also [relevant discussion in chat](https://chat.google.com/room/AAAATnZMo34/AXiRStLAO_4/AXiRStLAO_4?cls=10).

@rlazo Any thoughts?

Reply:
serialization is used here: https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/FirebaseDataConnect.kt#L26-L28

https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/FirebaseDataConnect.kt#L272-L278

coroutines is used here: https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/QuerySubscription.kt#L19

https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/QuerySubscription.kt#L52